src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -151,45 +151,66 @@ ExpressionEvaluator::ExpressionEvaluator(
 template <typename Result>
 std::shared_ptr<ExpressionResult<Result>> ExpressionEvaluator::evaluate_expression_to_result(
     const AbstractExpression& expression) {
+  // First, look in the cache
+  auto& evaluated_expression = _evaluated_expressions[expression.shared_from_this()];
```

Comment:
Sorry to be that guy, but I'd prefer using find+iterators here as to not leave the cache in a temporarily invalid state.